Synology exit-node crashes whole tailscaled

I installed Tailscale from the official Synology package manager on my Synology DS118. The default configuration seems to work. I tried to enable exit node on the Synology and it works for a little while. I confirmed this using whatsmyip on the other client. The traffic is routed through the Synology. However in the few moments it stops working.

Status after it stops working
`sudo /var/packages/Tailscale/target/bin/tailscale status 

Get "http://local-tailscaled.sock/localapi/v0/status": dial unix /var/packages/Tailscale/var/tailscaled.sock: connect: connection refused 
 `

Here is how I enable the exit node - using the Synology task setup after the boot up
`sleep 60 

/var/packages/Tailscale/target/bin/tailscale up --reset --advertise-exit-node 
`

Here are the hopefully relevant logs
`acceptUDP: could not create endpoint: no route 

acceptUDP: could not create endpoint: no route 

acceptUDP: could not create endpoint: no route 

acceptUDP: could not create endpoint: no route 

acceptUDP: could not create endpoint: no route 

[RATELIMIT] format("acceptUDP: could not create endpoint: %v") 

netstack: could not connect to local server at [2600:1901:0:c584::]:443: dial tcp [2600:1901:0:c584::]:443: connect: network is unreachable 

netstack: could not connect to local server at [2600:1901:0:c584::]:443: dial tcp [2600:1901:0:c584::]:443: connect: network is unreachable 

[RATELIMIT] format("acceptUDP: could not create endpoint: %v") (11 dropped) 

acceptUDP: could not create endpoint: no route 

acceptUDP: could not create endpoint: no route 

netstack: could not connect to local server at obfuscated:8080: operation was canceled 

netstack: could not connect to local server at obfuscated:8080: operation was canceled 

netstack: could not connect to local server at obfuscated:8080: operation was canceled 

netstack: could not connect to local server at obfuscated:8080: operation was canceled 

[RATELIMIT] format("netstack: could not connect to local server at %s: %v") 

panic: interface conversion: net.Addr is nil, not *net.TCPAddr 

  

goroutine 1123 [running]: 

tailscale.com/wgengine/netstack.(*Impl).forwardTCP(0x400017a3f0, 0x4000073c80, 0x40001a0510, 0x400069e21c, 0x4, 0x1f90) 

    tailscale.com@v1.1.1-0.20210610140605-f944614c5c0a/wgengine/netstack/netstack.go:491 +0xa18 

tailscale.com/wgengine/netstack.(*Impl).acceptTCP(0x400017a3f0, 0x4000016e80) 

    tailscale.com@v1.1.1-0.20210610140605-f944614c5c0a/wgengine/netstack/netstack.go:459 +0x278 

created by inet.af/netstack/tcpip/transport/tcp.(*Forwarder).HandlePacket 

    inet.af/netstack@v0.0.0-20210317161235-a1bf4e56ef22/tcpip/transport/tcp/forwarder.go:94 +0x288 
`

```
panic: interface conversion: net.Addr is nil, not *net.TCPAddr
goroutine 1123 [running]:
tailscale.com/wgengine/netstack.(*Impl).forwardTCP(0x400017a3f0, 0x4000073c80, 0x40001a0510, 0x400069e21c, 0x4, 0x1f90)
tailscale.com@v1.1.1-0.20210610140605-f944614c5c0a/wgengine/netstack/netstack.go:491 +0xa18 
```

I think I've got the commit correct for the 1.9.156 version in the Package Center:
https://github.com/tailscale/tailscale/blob/f944614c5c0a6fa9ef917dada05e1ca3e4478ac0/wgengine/netstack/netstack.go#L491

```
	backendLocalAddr := server.LocalAddr().(*net.TCPAddr)
	backendLocalIPPort, _ := netaddr.FromStdAddr(backendLocalAddr.IP, backendLocalAddr.Port, backendLocalAddr.Zone)
	clientRemoteIP, _ := netaddr.FromStdIP(client.RemoteAddr().(*net.TCPAddr).IP)         <-- HERE
	ns.e.RegisterIPPortIdentity(backendLocalIPPort, clientRemoteIP)
	defer ns.e.UnregisterIPPortIdentity(backendLocalIPPort)
```

That specific line in the current tree is https://github.com/tailscale/tailscale/blob/39610aeb09098142fdfd93a472a339ec33b24671/wgengine/netstack/netstack.go#L533, and there have been no changes in forwardTCP. There have been other changes elsewhere which might impact the issue since 1.9.156.

--- a/wgengine/netstack/netstack.go
+++ b/wgengine/netstack/netstack.go
@@ -468,19 +468,37 @@ func (ns *Impl) injectInbound(p *packet.Parsed, t *tstun.Wrapper) filter.Respons
 	return filter.DropSilently
 }
 
+func netaddrIPFromNetstackIP(s tcpip.Address) netaddr.IP {
+	switch len(s) {
+	case 4:
+		return netaddr.IPv4(s[0], s[1], s[2], s[3])
+	case 16:
+		var a [16]byte
+		copy(a[:], s)
+		return netaddr.IPFrom16(a)
+	}
+	return netaddr.IP{}
+}
+
 func (ns *Impl) acceptTCP(r *tcp.ForwarderRequest) {
 	reqDetails := r.ID()
 	if debugNetstack {
 		ns.logf("[v2] TCP ForwarderRequest: %s", stringifyTEI(reqDetails))
 	}
-	dialAddr := reqDetails.LocalAddress
-	dialNetAddr, _ := netaddr.FromStdIP(net.IP(dialAddr))
-	isTailscaleIP := tsaddr.IsTailscaleIP(dialNetAddr)
+	clientRemoteIP := netaddrIPFromNetstackIP(reqDetails.RemoteAddress)
+	if !clientRemoteIP.IsValid() {
+		ns.logf("invalid RemoteAddress in TCP ForwarderRequest: %s", stringifyTEI(reqDetails))
+		r.Complete(true)
+		return
+	}
+
+	dialIP := netaddrIPFromNetstackIP(reqDetails.LocalAddress)
+	isTailscaleIP := tsaddr.IsTailscaleIP(dialIP)
 	defer func() {
 		if !isTailscaleIP {
 			// if this is a subnet IP, we added this in before the TCP handshake
 			// so netstack is happy TCP-handshaking as a subnet IP
-			ns.removeSubnetAddress(dialNetAddr)
+			ns.removeSubnetAddress(dialIP)
 		}
 	}()
 	var wq waiter.Queue
@@ -490,21 +508,31 @@ func (ns *Impl) acceptTCP(r *tcp.ForwarderRequest) {
 		return
 	}
 	r.Complete(false)
+
+	// Asynchronously start the TCP handshake. Note that the
+	// gonet.TCPConn methods c.RemoteAddr() and c.LocalAddr() will
+	// return nil until the handshake actually completes. But we
+	// have the remote address in reqDetails instead, so we don't
+	// use RemoteAddr. The byte copies in both directions in
+	// forwardTCP will block until the TCP handshake is complete.
 	c := gonet.NewTCPConn(&wq, ep)
+
 	if ns.ForwardTCPIn != nil {
 		ns.ForwardTCPIn(c, reqDetails.LocalPort)
 		return
 	}
 	if isTailscaleIP {
-		dialAddr = tcpip.Address(net.ParseIP("127.0.0.1")).To4()
+		dialIP = netaddr.IPv4(127, 0, 0, 1)
 	}
-	ns.forwardTCP(c, &wq, dialAddr, reqDetails.LocalPort)
+	dialAddr := netaddr.IPPortFrom(dialIP, uint16(reqDetails.LocalPort))
+	ns.forwardTCP(c, clientRemoteIP, &wq, dialAddr)
 }
 
-func (ns *Impl) forwardTCP(client *gonet.TCPConn, wq *waiter.Queue, dialAddr tcpip.Address, dialPort uint16) {
+func (ns *Impl) forwardTCP(client *gonet.TCPConn, clientRemoteIP netaddr.IP, wq *waiter.Queue, dialAddr netaddr.IPPort) {
 	defer client.Close()
-	dialAddrStr := net.JoinHostPort(dialAddr.String(), strconv.Itoa(int(dialPort)))
+	dialAddrStr := dialAddr.String()
 	ns.logf("[v2] netstack: forwarding incoming connection to %s", dialAddrStr)
+
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 	waitEntry, notifyCh := waiter.NewChannelEntry(nil)
@@ -530,7 +558,6 @@ func (ns *Impl) forwardTCP(client *gonet.TCPConn, wq *waiter.Queue, dialAddr tcp
 	defer server.Close()
 	backendLocalAddr := server.LocalAddr().(*net.TCPAddr)
 	backendLocalIPPort, _ := netaddr.FromStdAddr(backendLocalAddr.IP, backendLocalAddr.Port, backendLocalAddr.Zone)
-	clientRemoteIP, _ := netaddr.FromStdIP(client.RemoteAddr().(*net.TCPAddr).IP)
 	ns.e.RegisterIPPortIdentity(backendLocalIPPort, clientRemoteIP)
 	defer ns.e.UnregisterIPPortIdentity(backendLocalIPPort)
 	connClosed := make(chan error, 2)

